client/src/controllers/RevealController.ts

```diff
@@ -0,0 +1,268 @@
+import { Controller } from '@hotwired/stimulus';
+
+/**
+ * Adds the ability to make the controlled element be used as an
+ * opening/closing (aka collapsing) element.
+ * Supports, and relies on, correct usage of aria-* attributes.
+ *
+ * @see https://w3c.github.io/aria/#aria-expanded
+ *
+ * @example
+ * <section data-controller="w-reveal">
+ *  <button type="button" data-action="w-reveal#toggle" data-w-reveal-target="toggle" aria-controls="my-content" type="button">Toggle</button>
+ *  <div id="my-content">CONTENT</div>
+ * </section>
+ */
+export class RevealController extends Controller<HTMLElement> {
+  static classes = [
+    'closed',
+    'closeIcon',
+    'initial',
+    'opened',
+    'openedContent',
+    'openIcon',
+  ];
+
+  static targets = ['content', 'toggle'];
+
+  static values = {
+    closed: { default: false, type: Boolean },
+    peeking: { default: false, type: Boolean },
+    peekTarget: { default: '', type: String },
+  };
+
+  declare closedValue: boolean;
+  declare peekingValue: boolean;
+
+  declare readonly closedClass: string;
+  declare readonly closedClasses: string[];
+  declare readonly closeIconClass: string;
+  declare readonly contentTarget: HTMLElement;
+  declare readonly contentTargets: HTMLElement[];
+  declare readonly hasClosedClass: boolean;
+  declare readonly hasCloseIconClass: string;
+  declare readonly hasContentTarget: boolean;
+  declare readonly hasOpenIconClass: string;
+  declare readonly hasToggleTarget: boolean;
+  declare readonly initialClasses: string[];
+  declare readonly openedClasses: string[];
+  declare readonly openedContentClasses: string[];
+  declare readonly openIconClass: string;
+  declare readonly peekTargetValue: string;
+  declare readonly toggleTarget: HTMLButtonElement;
+  declare readonly toggleTargets: HTMLButtonElement[];
+
+  cleanUpPeekListener?: () => void;
+
+  /**
+   * Ensure that targets are correctly set up, as this runs
+   * before the initial call to `closedValueChanged`.
+   * If there is no content element set, try to set one
+   * via `aria-controls` on the toggle, this also ensures
+   * that the toggleTarget has been set before the controller
+   * connects.
+   */
+  initialize() {
+    const toggleTarget = this.toggleTarget;
+
+    if (!this.hasContentTarget) {
```

Comment:
We probably should use it in the breadcrumbs but I didn't want to add too much to the html. We would need to add an ID to each Breadcrumbs item and then also add the set of the IDs to the toggle. 

But yes, it's not used by the breadcrumbs, but to add that in another PR seemed quite noisy. So I opted to keep it in.

Reply:
ok, I have cleaned this up further, sorry for the runaround, new update coming.

Reply:
Unfortunately, adding an ID to each breacrumb item is not practical without a larger change to the breadcrumbs include/shared template usage as we would probably need a base ID to build on.

Reply:
Thank you!

> We probably should use it in the breadcrumbs but I didn't want to add too much to the html. We would need to add an ID to each Breadcrumbs item and then also add the set of the IDs to the toggle.

> Unfortunately, adding an ID to each breacrumb item is not practical without a larger change to the breadcrumbs include/shared template usage as we would probably need a base ID to build on.

_That_ is exactly why I much preferred _not_ having the code to handle `aria-controls` just yet. The breadcrumbs HTML code in this PR works without changing much of the HTML (just sprinkling some Stimulus attributes is enough), so I didn't see why we need to add the `aria-controls` assumption for the toggles in the controller if it isn't used.

And I wasn't expecting you to add the `aria-controls` to the HTML so that the code is used, but rather: we stick to Stimulus attributes for now, and see if we can reintroduce the `aria-controls` handling later when there's a more appropriate use case, e.g. the panels where they already have `aria-controls` and `id`s and it's much simpler to use those there instead of adding Stimulus attributes.

Sorry if I wasn't clear, glad you didn't proceed with it 😅